### PR TITLE
Nomad HAProxy load balancing guide

### PR DESCRIPTION
**Pull request by Omar-Khawaja**
Merge in [NGINX load balancing PR](https://github.com/hashicorp/nomad/pull/6529) first

### CODE REVIEW

**Comment by shantanugadgil** [comment]
Kudos on this and the NGINX LB docs. Many folks are looking for quickstart guides and this helps to familiarize them quickly.

File: website/source/guides/load-balancing/haproxy.html.md [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+
+## Prerequisites
+
+To perform the tasks described in this guide, you need to have a Nomad
+environment with Consul installed. You can use this [repo][terraform-repo] to
+easily provision a sandbox environment. This guide will assume a cluster with
+one server node and three client nodes.
+
+-> **Note:** This guide is for demo purposes and is only using a single
+server node. In a production cluster, 3 or 5 server nodes are recommended.
```
[the remark this answers is not in the document]
Comment (cgbaker) [reply website/source/guides/load-balancing/haproxy.html.md]:
```suggestion
-> **Note:** This guide is for demo purposes and only assumes a single server node. Please consult the [reference architecture](/guides/install/production/reference-architecture.html#high-availability) for production configuration.
```

File: website/source/guides/load-balancing/haproxy.html.md [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+
+## Steps
+
+### Step 1: Create a Job for Demo Web App
+
+Create a job for a demo web application and name the file 'webapp.nomad'
```
Comment (cgbaker) [inline website/source/guides/load-balancing/haproxy.html.md]:
```suggestion
Create a job for a demo web application and name the file 'webapp.nomad':
```

File: website/source/guides/load-balancing/haproxy.html.md [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+      driver = "docker"
+
+      config {
+        image = "hashicorp/demo-webapp-lb-guide"
+      }
+
+      resources {
+        network {
+          mbits = 10
+          port  "http"{}
```
Comment (cgbaker) [inline website/source/guides/load-balancing/haproxy.html.md]:
```suggestion
          port  "http" {}
```

Comment (cgbaker) [reply website/source/guides/load-balancing/haproxy.html.md]: weird, it looks like `hclfmt` prefers no space.... do with that what you will.

File: website/source/guides/load-balancing/haproxy.html.md [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+defined:
+
+- The `backend http_back` stanza in the HAProxy config includes two main
+  settings: `balance type` and `server-template`. The `balance type` is round
+  robin and will load balance across the available service in order.
+- The `server-template` option allows Consul service registrations to configure
+  HAProxy's backend server pool. Because of this, you do not need to explicitly
+  add your backend servers' IP addresses. We have specified a server-template
+  named mywebapp. The template name is not tied to the service name which is
+  registered in Consul.
```
Comment (cgbaker) [inline website/source/guides/load-balancing/haproxy.html.md]: I'd break the `balance type` and `server-template` discussions into dedicated bullets... or maybe even "demote" the first bullet to paragraph text, putting the two explanations into their own list.

File: website/source/guides/load-balancing/haproxy.html.md [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+- _demo-webapp._tcp.service.consul allows HAProxy to use the DNS SRV record for
```
Comment (cgbaker) [inline website/source/guides/load-balancing/haproxy.html.md]:
```suggestion
- `_demo-webapp._tcp.service.consul` allows HAProxy to use the DNS SRV record for
```

File: website/source/guides/load-balancing/haproxy.html.md [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+  the backend service demo-webapp.service.consul to discover the available
```
Comment (cgbaker) [inline website/source/guides/load-balancing/haproxy.html.md]:
```suggestion
  the backend service `demo-webapp.service.consul` to discover the available
```

File: website/source/guides/load-balancing/haproxy.html.md [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+  instances of the service.
+
+Additionally, keep in mind the following points from the Nomad job spec:
+
+- We have statically set the port of our load balancer to `8080`. This will
+  allow us to query `haproxy.service.consul:8080` from anywhere inside our cluster
+  so we can reach our web application.
+- Please note that although we have defined the template [inline][inline], we
+  can use the template stanza [in conjunction with the artifact
```
Comment (cgbaker) [inline website/source/guides/load-balancing/haproxy.html.md]:
```suggestion
  could alternatively use the template stanza [in conjunction with the artifact
```

File: website/source/guides/load-balancing/haproxy.html.md [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+    Evaluation within deployment: "e8214434"
+    Allocation "53145b8b" created: node "d16a11fb", group "haproxy"
+    Evaluation status changed: "pending" -> "complete"
+==> Evaluation "937b1a2d" finished with status "complete"
+```
+
+### Step 5: Check the HAProxy Statistics Page
+
+You can visit the statistics and monitoring page for HAProxy at
+http://`<Your-HAProxy-IP-address>:1936`. You can use this page to verify your
```
Comment (cgbaker) [inline website/source/guides/load-balancing/haproxy.html.md]:
```suggestion
`http://<Your-HAProxy-IP-address>:1936`. You can use this page to verify your
```

File: website/source/guides/load-balancing/haproxy.html.md [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+    task "haproxy" {
+      driver = "docker"
+
+      config {
+        image        = "haproxy:2.0"
+        network_mode = "host"
+
+        port_map {
+          http = 8080
+        }
```
Comment (cgbaker) [inline website/source/guides/load-balancing/haproxy.html.md]: it it worthwhile/necessary to add 1936 to this and the resources stanza below?

File: website/source/guides/load-balancing/haproxy.html.md [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+### Step 5: Check the HAProxy Statistics Page
+
+You can visit the statistics and monitoring page for HAProxy at
+http://`<Your-HAProxy-IP-address>:1936`. You can use this page to verify your
+settings and for basic monitoring.
+
+[![Home Page][haproxy_ui]][haproxy_ui]
+
+Notice there are 10 pre-provisioned load balancer backend slots for your service
+but that only three of them are being used.
```
Comment (cgbaker) [inline website/source/guides/load-balancing/haproxy.html.md]:
```suggestion
but that only three of them are being used, corresponding to the three allocations in the current job.
```

File: website/source/guides/load-balancing/haproxy.html.md [hunk 11]
```diff
@@ -0,0 +1,272 @@
...
+
+### Step 6: Make a Request to the Load Balancer
+
+If you query the HAProxy load balancer, you should be able to see a response
+similar to the one shown below (keep in mind you need to run this command from a
```
Comment (cgbaker) [inline website/source/guides/load-balancing/haproxy.html.md]:
```suggestion
similar to the one shown below (this command should be run from a
```

File: website/source/guides/load-balancing/haproxy.html.md [hunk 12]
```diff
@@ -0,0 +1,272 @@
...
+Welcome! You are on node 172.31.54.242:20124
+```
+
+Note that your request has been forwarded to one of the several deployed
+instances of the demo web application (which is spread across 3 Nomad clients).
+The output shows the IP address of the host it is deployed on. If you repeat
+your requests, you will see that the IP address changes.
+
+* Note: if you would like to access HAProxy from outside your cluster, you
+  should set up a load balancer in your environment that maps to an active port
```
Comment (cgbaker) [inline website/source/guides/load-balancing/haproxy.html.md]:
```suggestion
  can set up a load balancer in your environment that maps to an active port
```

File: website/source/guides/load-balancing/haproxy.html.md (hunk 9 above)
Comment (Omar-Khawaja) [reply website/source/guides/load-balancing/haproxy.html.md]: Just added this for clarity. Thanks.

**Pull request merged**